File: packages/shared/src/agent/knowledge-search.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const DEFAULT_ROOT = path.resolve(__dirname, '..', '..', '..', '..');
const MAX_QUERY_CHARS = 8_000;
const MAX_FILES = 80;
const MAX_FILE_BYTES = 256 * 1024;
const MAX_TOTAL_BYTES = 2 * 1024 * 1024;
const MAX_RESULTS = 8;
const ALLOWED_EXTENSIONS = new Set(['.md', '.mdx', '.txt']);
const STOP_WORDS = new Set([
  'about', 'after', 'also', 'and', 'are', 'can', 'check', 'find', 'for', 'from', 'have',
  'into', 'look', 'our', 'please', 'search', 'show', 'that', 'the', 'their', 'this', 'what',
  'when', 'where', 'which', 'with', 'workspace', 'would', 'your',
]);

class KnowledgeSearchError extends Error {
  constructor(message, status = 400) {
    super(message);
    this.name = 'KnowledgeSearchError';
    this.status = status;
  }
}

function normalizeQuery(value) {
  if (typeof value !== 'string') throw new KnowledgeSearchError('query must be a string.');
  const query = value.replace(/\s+/g, ' ').trim();
  if (!query) throw new KnowledgeSearchError('Describe what you want to find.');
  if (query.length > MAX_QUERY_CHARS) throw new KnowledgeSearchError(`query must be ${MAX_QUERY_CHARS.toLocaleString()} characters or fewer.`);
  return query;
}

function queryTerms(query) {
  return [...new Set(query.toLowerCase().match(/[a-z0-9][a-z0-9_-]{1,}/g) || [])]
    .filter((term) => term.length > 2 && !STOP_WORDS.has(term))
    .slice(0, 20);
}

function collectDocumentPaths(root) {
  const resolvedRoot = path.resolve(root);
  const files = [];
  const addFile = (candidate) => {
    if (files.length >= MAX_FILES) return;
    let stat;
    try {
      stat = fs.lstatSync(candidate);
    } catch (_) {
      return;
    }
    if (!stat.isFile() || stat.isSymbolicLink() || stat.size > MAX_FILE_BYTES) return;
    if (!ALLOWED_EXTENSIONS.has(path.extname(candidate).toLowerCase())) return;
    const resolved = path.resolve(candidate);
    if (resolved !== resolvedRoot && !resolved.startsWith(`${resolvedRoot}${path.sep}`)) return;
    files.push({ absolute: resolved, relative: path.relative(resolvedRoot, resolved), size: stat.size });
  };
  addFile(path.join(resolvedRoot, 'README.md'));

  const walk = (directory, depth) => {
    if (depth > 3 || files.length >= MAX_FILES) return;
    let entries;
    try {
      entries = fs.readdirSync(directory, { withFileTypes: true });
    } catch (_) {
      return;
    }
    for (const entry of entries) {
      if (files.length >= MAX_FILES) break;
      if (entry.isSymbolicLink()) continue;
      const candidate = path.join(directory, entry.name);
      if (entry.isDirectory()) walk(candidate, depth + 1);
      else if (entry.isFile()) addFile(candidate);
    }
  };
  walk(path.join(resolvedRoot, 'docs'), 0);
  return files;
}

function documentTitle(lines, relative) {
  const heading = lines.find((line) => /^#{1,3}\s+\S/.test(line));
  return heading ? heading.replace(/^#{1,3}\s+/, '').trim().slice(0, 160) : path.basename(relative, path.extname(relative));
}

function cleanSnippet(lines, index) {
  return lines
    .slice(Math.max(0, index - 1), Math.min(lines.length, index + 2))
    .map((line) => line.replace(/^#{1,6}\s+/, '').replace(/\s+/g, ' ').trim())
    .filter(Boolean)
    .join(' ')
    .slice(0, 520);
}

function scoreLine(line, terms, phrase) {
  const lower = line.toLowerCase();
  const termScore = terms.reduce((score, term) => score + (lower.includes(term) ? 2 : 0), 0);
  return termScore + (phrase.length >= 5 && lower.includes(phrase) ? 5 : 0);
}

/** Bounded lexical search over reviewed workspace documentation only. */
function searchDocuments(value, options = {}) {
  const query = normalizeQuery(value);
  const terms = queryTerms(query);
  if (!terms.length) return { query, indexedFiles: 0, results: [] };
  const paths = collectDocumentPaths(options.root || DEFAULT_ROOT);
  let totalBytes = 0;
  const matches = [];
  for (const file of paths) {
    if (totalBytes + file.size > MAX_TOTAL_BYTES) break;
    totalBytes += file.size;
    let text;
    try {
      text = fs.readFileSync(file.absolute, 'utf8');
    } catch (_) {
      continue;
    }
    const lines = text.split(/\r?\n/);
    let best = null;
    for (let index = 0; index < lines.length; index += 1) {
      const score = scoreLine(lines[index], terms, query.toLowerCase());
      if (score > 0 && (!best || score > best.score)) best = { score, index };
    }
    if (!best) continue;
    matches.push({
      type: 'Workspace document',
      title: documentTitle(lines, file.relative),
      path: file.relative.split(path.sep).join('/'),
      snippet: cleanSnippet(lines, best.index),
      score: best.score,
    });
  }
  const limit = Math.min(MAX_RESULTS, Math.max(1, Number(options.limit) || MAX_RESULTS));
  return {
    query,
    indexedFiles: paths.length,
    results: matches.sort((a, b) => b.score - a.score || a.path.localeCompare(b.path)).slice(0, limit),
  };
}

module.exports = {
  DEFAULT_ROOT,
  MAX_FILES,
  MAX_FILE_BYTES,
  MAX_TOTAL_BYTES,
  KnowledgeSearchError,
  normalizeQuery,
  collectDocumentPaths,
  searchDocuments,
};

File: ai_fleet/agent/knowledge_search.py
"""Bounded lexical search over reviewed workspace documentation (port of
agent/knowledge-search.js).

Security-critical constraints (preserved exactly from the JS): only README.md +
docs/ are indexed, symlinks are refused, resolved paths must stay inside the
root (no escape via ``..``/links), only a small allow-list of extensions is read,
and hard caps bound the number/size of files scanned and the walk depth.
"""

from __future__ import annotations

import os
import re
import stat as stat_mod
from pathlib import Path

# The JS resolves 4 levels up from __dirname (packages/shared/src/agent) = repo
# root. Here the module lives at ai_fleet/agent/knowledge_search.py, so parents[2]
# (agent -> ai_fleet -> repo root) is the repo root.
DEFAULT_ROOT = str(Path(__file__).resolve().parents[2])
MAX_QUERY_CHARS = 8_000
MAX_FILES = 80
MAX_FILE_BYTES = 256 * 1024
MAX_TOTAL_BYTES = 2 * 1024 * 1024
MAX_RESULTS = 8
ALLOWED_EXTENSIONS = frozenset({".md", ".mdx", ".txt"})
STOP_WORDS = frozenset({
    "about", "after", "also", "and", "are", "can", "check", "find", "for", "from", "have",
    "into", "look", "our", "please", "search", "show", "that", "the", "their", "this", "what",
    "when", "where", "which", "with", "workspace", "would", "your",
})

_WS_RE = re.compile(r"\s+")
_TERM_RE = re.compile(r"[a-z0-9][a-z0-9_-]{1,}")
_HEADING_RE = re.compile(r"^#{1,3}\s+\S")
_HEADING_STRIP_RE = re.compile(r"^#{1,3}\s+")
_HEADING_STRIP6_RE = re.compile(r"^#{1,6}\s+")
_NEWLINE_RE = re.compile(r"\r?\n")


class KnowledgeSearchError(Exception):
    def __init__(self, message, status: int = 400):
        super().__init__(message)
        self.name = "KnowledgeSearchError"
        self.message = message
        self.status = status


def normalize_query(value):
    if not isinstance(value, str):
        raise KnowledgeSearchError("query must be a string.")
    query = _WS_RE.sub(" ", value).strip()
    if not query:
        raise KnowledgeSearchError("Describe what you want to find.")
    if len(query) > MAX_QUERY_CHARS:
        raise KnowledgeSearchError(f"query must be {MAX_QUERY_CHARS:,} characters or fewer.")
    return query


def query_terms(query):
    # [...new Set(...)] dedupes preserving first-seen order.
    unique = list(dict.fromkeys(_TERM_RE.findall(query.lower())))
    return [term for term in unique if len(term) > 2 and term not in STOP_WORDS][:20]


def collect_document_paths(root):
    resolved_root = str(Path(root).resolve())
    files: list[dict] = []

    def add_file(candidate: str) -> None:
        if len(files) >= MAX_FILES:
            return
        try:
            st = os.lstat(candidate)
        except OSError:
            return
        is_file = stat_mod.S_ISREG(st.st_mode)
        is_symlink = stat_mod.S_ISLNK(st.st_mode)
        if (not is_file) or is_symlink or st.st_size > MAX_FILE_BYTES:
            return
        if Path(candidate).suffix.lower() not in ALLOWED_EXTENSIONS:
            return
        resolved = str(Path(candidate).resolve())
        if resolved != resolved_root and not resolved.startswith(f"{resolved_root}{os.sep}"):
            return
        files.append({
            "absolute": resolved,
            "relative": os.path.relpath(resolved, resolved_root),
            "size": st.st_size,
        })

    add_file(os.path.join(resolved_root, "README.md"))

    def walk(directory: str, depth: int) -> None:
        if depth > 3 or len(files) >= MAX_FILES:
            return
        try:
            entries = list(os.scandir(directory))
        except OSError:
            return
        for entry in entries:
            if len(files) >= MAX_FILES:
                break
            if entry.is_symlink():
                continue
            candidate = os.path.join(directory, entry.name)
            if entry.is_dir(follow_symlinks=False):
                walk(candidate, depth + 1)
            elif entry.is_file(follow_symlinks=False):
                add_file(candidate)

    walk(os.path.join(resolved_root, "docs"), 0)
    return files


def document_title(lines, relative):
    heading = next((line for line in lines if _HEADING_RE.match(line)), None)
    if heading:
        return _HEADING_STRIP_RE.sub("", heading).strip()[:160]
    return Path(relative).stem


def clean_snippet(lines, index):
    segment = lines[max(0, index - 1): min(len(lines), index + 2)]
    cleaned = []
    for line in segment:
        text = _HEADING_STRIP6_RE.sub("", line)
        text = _WS_RE.sub(" ", text).strip()
        if text:
            cleaned.append(text)
    return " ".join(cleaned)[:520]


def score_line(line, terms, phrase):
    lower = line.lower()
    term_score = sum(2 for term in terms if term in lower)
    phrase_score = 5 if len(phrase) >= 5 and phrase in lower else 0
    return term_score + phrase_score


def _to_number(value):
    # Mirror JS Number(x): non-numeric -> NaN, which is falsy in `x || default`.
    try:
        n = float(value)
    except (TypeError, ValueError):
        return None
    if n != n:  # NaN
        return None
    return n


def search_documents(value, options=None):
    """Bounded lexical search over reviewed workspace documentation only."""
    options = options or {}
    query = normalize_query(value)
    terms = query_terms(query)
    if not terms:
        return {"query": query, "indexedFiles": 0, "results": []}
    paths = collect_document_paths(options.get("root") or DEFAULT_ROOT)
    phrase = query.lower()
    total_bytes = 0
    matches = []
    for file in paths:
        if total_bytes + file["size"] > MAX_TOTAL_BYTES:
            break
        total_bytes += file["size"]
        try:
            with open(file["absolute"], "r", encoding="utf-8", errors="replace") as fh:
                text = fh.read()
        except Exception:  # noqa: BLE001 — JS `catch (_) { continue; }`
            continue
        lines = _NEWLINE_RE.split(text)
        best = None
        for index, line in enumerate(lines):
            score = score_line(line, terms, phrase)
            if score > 0 and (best is None or score > best["score"]):
                best = {"score": score, "index": index}
        if best is None:
            continue
        matches.append({
            "type": "Workspace document",
            "title": document_title(lines, file["relative"]),
            "path": file["relative"].replace(os.sep, "/"),
            "snippet": clean_snippet(lines, best["index"]),
            "score": best["score"],
        })
    raw_limit = _to_number(options.get("limit")) or MAX_RESULTS
    limit = int(min(MAX_RESULTS, max(1, raw_limit)))
    ordered = sorted(matches, key=lambda m: (-m["score"], m["path"]))
    return {"query": query, "indexedFiles": len(paths), "results": ordered[:limit]}
